Sources/FoundationEssentials/Formatting/Date+ISO8601FormatStyle.swift

```diff
@@ -0,0 +1,487 @@
+        new.timeZoneSeparator = separator
+        return new
+    }
+
+    public func dateSeparator(_ separator: DateSeparator) -> Self {
+        var new = self
+        new.dateSeparator = separator
+        return new
+    }
+
+    public func dateTimeSeparator(_ separator: DateTimeSeparator) -> Self {
+        var new = self
+        new.dateTimeSeparator = separator
+        return new
+    }
+
+    public func timeSeparator(_ separator: TimeSeparator) -> Self {
+        var new = self
+        new.timeSeparator = separator
+        return new
+    }
+
+    public func timeZoneSeparator(_ separator: TimeZoneSeparator) -> Self {
+        var new = self
+        new.timeZoneSeparator = separator
+        return new
+    }
+}
+
+#if FOUNDATION_FRAMEWORK
+@available(macOS 12.0, iOS 15.0, tvOS 15.0, watchOS 8.0, *)
+extension ISO8601DateFormatter.Options : Hashable {}
+#endif // FOUNDATION_FRAMEWORK
+
+@available(macOS 12.0, iOS 15.0, tvOS 15.0, watchOS 8.0, *)
+extension Date.ISO8601FormatStyle : FormatStyle {
+    public func format(_ value: Date) -> String {
+        var whichComponents = Calendar.ComponentSet()
+        let fields = formatFields
+        
+        // If we use week of year, don't bother with year
+        if fields.contains(.year) && !fields.contains(.weekOfYear) {
+            whichComponents.insert(.year)
+        }
+        
+        if fields.contains(.month) {
+            whichComponents.insert(.month)
+        }
+        
+        if fields.contains(.weekOfYear) {
+            whichComponents.insert([.weekOfYear, .yearForWeekOfYear])
+        }
+        
+        if fields.contains(.day) {
+            if fields.contains(.weekOfYear) {
+                whichComponents.insert(.weekday)
+            } else if fields.contains(.month) {
+                whichComponents.insert(.day)
+            } else {
+                whichComponents.insert(.dayOfYear)
+            }
+        }
+        
+        if fields.contains(.time) {
+            whichComponents.insert([.hour, .minute, .second])
+            if includingFractionalSeconds {
+                whichComponents.insert(.nanosecond)
+            }
+        }
+
+        let components = _calendar.dateComponents(whichComponents, from: value)
+        var needSeparator = false
+        let capacity = 128 // It is believed no ISO8601 date can exceed this size
+        let result = withUnsafeTemporaryAllocation(of: CChar.self, capacity: capacity + 1) { _buffer in
+            var buffer = OutputBuffer(initializing: _buffer.baseAddress!, capacity: _buffer.count)
+            
+            let asciiZero = CChar(48)
+            
+            func append(_ i: Int, zeroPad: Int, buffer: inout OutputBuffer<CChar>) {
+                if i < 10 {
+                    if zeroPad - 1 > 0 {
+                        for _ in 0..<zeroPad-1 { buffer.appendElement(asciiZero) }
+                    }
+                    buffer.appendElement(CChar(i + 48))
```

Comment:
Maybe we could use the `asciiZero` value instead of `48` here and below

Reply:
Yah, I can fix that in a follow-up PR.